/*
 * Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, and/or sell copies of the Software, and to permit persons
 * to whom the Software is furnished to do so, provided that the above
 * copyright notice(s) and this permission notice appear in all copies of
 * the Software and that both the above copyright notice(s) and this
 * permission notice appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
 * OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
 * INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
 * FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Except as contained in this notice, the name of a copyright holder
 * shall not be used in advertising or otherwise to promote the sale, use
 * or other dealings in this Software without prior written authorization
 * of the copyright holder.
 *
 */

Sun bug 5077981/GNOME bug 147579:
	There should be an option to extend/disable lockout timer
	http://bugs.opensolaris.org/bugdatabase/view_bug.do?bug_id=5077981
	http://bugzilla.gnome.org/show_bug.cgi?id=147579

diff -urp -x '*~' driver/XScreenSaver.ad.in driver/XScreenSaver.ad.in
--- driver/XScreenSaver.ad.in	2006-08-08 15:24:43.355144000 -0700
+++ driver/XScreenSaver.ad.in	2006-08-08 15:24:49.627321000 -0700
@@ -34,6 +34,7 @@
 *cycle:			0:10:00
 *lockTimeout:		0:30:00
 *passwdTimeout:		0:02:00
+*passwdTimeoutEnabled:	True
 *dpmsEnabled:		True
 *dpmsStandby:		0:24:00
 *dpmsSuspend:		0:27:00
diff -urp -x '*~' driver/demo-Gtk.c driver/demo-Gtk.c
--- driver/demo-Gtk.c	2006-08-08 15:24:43.446343000 -0700
+++ driver/demo-Gtk.c	2006-08-08 15:24:49.630059000 -0700
@@ -1500,6 +1500,9 @@ flush_dialog_changes_and_save (state *s)
 
   MINUTES  (&p2->timeout,         "timeout_spinbutton");
   MINUTES  (&p2->cycle,           "cycle_spinbutton");
+  /*bugid 5077981 pwd */
+  CHECKBOX (p2->pwd_p,           "pwd_button");
+  MINUTES  (&p2->passwd_timeout,    "pwd_spinbutton");
   CHECKBOX (p2->lock_p,           "lock_button");
   MINUTES  (&p2->lock_timeout,    "lock_spinbutton");
 
@@ -1591,6 +1594,9 @@ flush_dialog_changes_and_save (state *s)
   COPY(cycle,          "cycle");
   COPY(lock_p,         "lock_p");
   COPY(lock_timeout,   "lock_timeout");
+/*bugid 5077981 pwd */
+  COPY(pwd_p,         "pwd_p");
+  COPY(passwd_timeout,   "passwd_timeout");
 
   COPY(dpms_enabled_p, "dpms_enabled_p");
   COPY(dpms_standby,   "dpms_standby");
@@ -2706,6 +2712,9 @@ populate_prefs_page (state *s)
   FMT_MINUTES ("timeout_spinbutton",      p->timeout);
   FMT_MINUTES ("cycle_spinbutton",        p->cycle);
   FMT_MINUTES ("lock_spinbutton",         p->lock_timeout);
+  /*bugid 5077981 pwd */
+  FMT_MINUTES ("pwd_spinbutton",         p->passwd_timeout);
+  FMT_MINUTES ("lock_spinbutton",         p->lock_timeout);
   FMT_MINUTES ("dpms_standby_spinbutton", p->dpms_standby);
   FMT_MINUTES ("dpms_suspend_spinbutton", p->dpms_suspend);
   FMT_MINUTES ("dpms_off_spinbutton",     p->dpms_off);
@@ -2717,6 +2726,10 @@ populate_prefs_page (state *s)
 # define TOGGLE_ACTIVE(NAME,ACTIVEP) \
   gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (name_to_widget (s,(NAME))),\
                                 (ACTIVEP))
+/*bugid 5077981 pwd */
+
+
+  TOGGLE_ACTIVE ("pwd_button",       p->pwd_p);
 
   TOGGLE_ACTIVE ("lock_button",       p->lock_p);
 #if 0
@@ -2822,6 +2835,10 @@ populate_prefs_page (state *s)
 
     /* Blanking and Locking
      */
+    /*bugid 5077081 pwd */
+    SENSITIZE ("pwd_spinbutton", p->pwd_p);
+    SENSITIZE ("pwd_mlabel",     p->pwd_p);
+
     SENSITIZE ("lock_button",     can_lock_p);
     SENSITIZE ("lock_spinbutton", can_lock_p && p->lock_p);
     SENSITIZE ("lock_mlabel",     can_lock_p && p->lock_p);
@@ -2990,10 +3007,12 @@ fix_text_entry_sizes (state *s)
 {
   GtkWidget *w;
 
+/*bugid 5077981 pwd */
+
 # if 0   /* appears no longer necessary with Gtk 1.2.10 */
   const char * const spinbuttons[] = {
     "timeout_spinbutton", "cycle_spinbutton", "lock_spinbutton",
-    "dpms_standby_spinbutton", "dpms_suspend_spinbutton",
+	"pwd_spinbutton", "dpms_standby_spinbutton", "dpms_suspend_spinbutton",
     "dpms_off_spinbutton",
     "-fade_spinbutton" };
   int i;
diff -urp -x '*~' driver/prefs.c driver/prefs.c
--- driver/prefs.c	2006-05-23 15:03:47.000000000 -0700
+++ driver/prefs.c	2006-08-08 15:24:49.631015000 -0700
@@ -243,6 +243,7 @@ get_byte_resource (Display *dpy, char *n
   return n;
 }
 
+/* *bugid 5077981 pwd timeout */
 
 static const char * const prefs[] = {
   "timeout",
@@ -251,6 +252,7 @@ static const char * const prefs[] = {
   "lockVTs",			/* not saved */
   "lockTimeout",
   "passwdTimeout",
+  "passwdTimeoutEnabled",
   "visualID",
   "installColormap",
   "verbose",
@@ -781,6 +783,9 @@ write_init_file (Display *dpy,
       CHECK("lockVTs")		continue;  /* don't save, unused */
       CHECK("lockTimeout")	type = pref_time, t = p->lock_timeout;
       CHECK("passwdTimeout")	type = pref_time, t = p->passwd_timeout;
+
+/* *bugid 5077981 pwd timeout */
+      CHECK("passwdTimeoutEnabled")		type = pref_bool, b = p->pwd_p;
       CHECK("visualID")		type = pref_str,  s =    visual_name;
       CHECK("installColormap")	type = pref_bool, b = p->install_cmap_p;
       CHECK("verbose")		type = pref_bool, b = p->verbose_p;
@@ -1062,6 +1067,8 @@ load_init_file (Display *dpy, saver_pref
   p->lock_timeout    = 1000 * get_minutes_resource (dpy, "lockTimeout", "Time");
   p->cycle           = 1000 * get_minutes_resource (dpy, "cycle", "Time");
   p->passwd_timeout  = 1000 * get_seconds_resource (dpy, "passwdTimeout", "Time");
+  /* *bugid 5077981 pwd timeout */
+  p->pwd_p	    = get_boolean_resource (dpy, "passwdTimeoutEnabled", "Boolean");
   p->pointer_timeout = 1000 * get_seconds_resource (dpy, "pointerPollTime", "Time");
   p->pointer_hysteresis = get_integer_resource (dpy, "pointerHysteresis","Integer");
   p->notice_events_timeout = 1000*get_seconds_resource(dpy,
diff -urp -x '*~' driver/prefs.h driver/prefs.h
--- driver/prefs.h	2006-05-17 16:35:54.000000000 -0700
+++ driver/prefs.h	2006-08-08 15:24:49.631360000 -0700
@@ -53,6 +53,8 @@ struct saver_preferences {
   Bool xsync_p;			/* whether XSynchronize has been called */
 
   Bool lock_p;			/* whether to lock as well as save */
+  Bool pwd_p;			/* whether to disable/enable pwd timeout */
+				/* bugid 5077981 */
 
   Bool fade_p;			/* whether to fade to black, if possible */
   Bool unfade_p;		/* whether to fade from black, if possible */
diff -urp -x '*~' driver/xscreensaver-demo.glade2 driver/xscreensaver-demo.glade2
--- driver/xscreensaver-demo.glade2	2005-08-09 01:37:05.000000000 -0700
+++ driver/xscreensaver-demo.glade2	2006-08-08 15:24:49.632892000 -0700
@@ -160,7 +160,7 @@
 		  <child>
 		    <widget class="GtkTable" id="blanking_table">
 		      <property name="visible">True</property>
-		      <property name="n_rows">3</property>
+		      <property name="n_rows">4</property>
 		      <property name="n_columns">4</property>
 		      <property name="homogeneous">False</property>
 		      <property name="row_spacing">2</property>
@@ -326,6 +326,65 @@
 		      </child>
 
 		      <child>
+			<widget class="GtkSpinButton" id="pwd_spinbutton">
+			  <property name="visible">True</property>
+			  <property name="tooltip" translatable="yes">How long the unlock dialog waits for input before disappearing.</property>
+			  <property name="can_focus">True</property>
+			  <property name="climb_rate">15</property>
+			  <property name="digits">0</property>
+			  <property name="numeric">True</property>
+			  <property name="update_policy">GTK_UPDATE_ALWAYS</property>
+			  <property name="snap_to_ticks">True</property>
+			  <property name="wrap">False</property>
+			  <property name="adjustment">0 0 720 1 15 15</property>
+			  <signal name="activate" handler="pref_changed_cb"/>
+			  <signal name="focus_out_event" handler="pref_changed_event_cb"/>
+			  <signal name="value_changed" handler="pref_changed_cb"/>
+			</widget>
+			<packing>
+			  <property name="left_attach">2</property>
+			  <property name="right_attach">3</property>
+			  <property name="top_attach">3</property>
+			  <property name="bottom_attach">4</property>
+			  <property name="y_padding">10</property>
+			  <property name="x_options">fill</property>
+			  <property name="y_options"></property>
+			</packing>
+		      </child>
+
+		      <child>
+			<widget class="GtkEventBox" id="pwd_button_eventbox">
+			  <property name="visible">True</property>
+			  <property name="tooltip" translatable="yes">Whether the unlock dialog box should disappear after a timeout.</property>
+			  <property name="visible_window">True</property>
+			  <property name="above_child">False</property>
+
+			  <child>
+			    <widget class="GtkCheckButton" id="pwd_button">
+			      <property name="visible">True</property>
+			      <property name="can_focus">True</property>
+			      <property name="label" translatable="yes">Timeout _Unlock After</property>
+			      <property name="use_underline">True</property>
+			      <property name="relief">GTK_RELIEF_NORMAL</property>
+			      <property name="focus_on_click">True</property>
+			      <property name="active">False</property>
+			      <property name="inconsistent">False</property>
+			      <property name="draw_indicator">True</property>
+			      <signal name="toggled" handler="pref_changed_cb"/>
+			    </widget>
+			  </child>
+			</widget>
+			<packing>
+			  <property name="left_attach">0</property>
+			  <property name="right_attach">2</property>
+			  <property name="top_attach">3</property>
+			  <property name="bottom_attach">4</property>
+			  <property name="x_options">fill</property>
+			  <property name="y_options">fill</property>
+			</packing>
+		      </child>
+
+		      <child>
 			<widget class="GtkLabel" id="cycle_mlabel">
 			  <property name="visible">True</property>
 			  <property name="label" translatable="yes">minutes</property>
@@ -372,6 +431,29 @@
 		      </child>
 
 		      <child>
+			<widget class="GtkLabel" id="pwd_mlabel">
+			  <property name="visible">True</property>
+			  <property name="label" translatable="yes">minutes</property>
+			  <property name="use_underline">False</property>
+			  <property name="use_markup">False</property>
+			  <property name="justify">GTK_JUSTIFY_LEFT</property>
+			  <property name="wrap">False</property>
+			  <property name="selectable">False</property>
+			  <property name="xalign">0</property>
+			  <property name="yalign">0.5</property>
+			  <property name="xpad">8</property>
+			  <property name="ypad">0</property>
+			</widget>
+			<packing>
+			  <property name="left_attach">3</property>
+			  <property name="right_attach">4</property>
+			  <property name="top_attach">3</property>
+			  <property name="bottom_attach">4</property>
+			  <property name="y_options"></property>
+			</packing>
+		      </child>
+
+		      <child>
 			<widget class="GtkLabel" id="timeout_label">
 			  <property name="visible">True</property>
 			  <property name="label" translatable="yes">_Blank After</property>
